### ISSUE TITLE
preload and this.error not working correctly

### ISSUE THREAD

**Issue by arxpoetica** [sveltejs/sapper#290]
Consider the following:

```html
<script>
    export default {
        async preload(req) {
            if (process.browser) {
                console.log('process.browser')
            } else {
                const { validate } = await import('../_services/auth-check')
                const test = validate(req)
                if (test.unauthorized) {
                    this.error(404, 'Not found')
                    return { nogo: true }
                }
            }
        },
    }
</script>
```

Note that the `return { nogo: true }` does run, but it's never serialized because of this line here: https://github.com/sveltejs/sapper/blob/master/src/middleware.ts#L255

In other words, any time there's an `if (process.browser) { ... }` in a `preload`, it will *always* run, even if the something is serialized and returned.

In both cases, running both `this.error...` from the server and something else in `if (process.browser) { ... }` has the unintended consequences of making the 404 page not render (it flickers momentarily, then vanishes). In other words, `process.browser` doesn't see anything in `preload` here: https://github.com/sveltejs/sapper/blob/master/src/middleware.ts#L393

Dunno what the intended behavior is, but this feels wrong. A related bug (should I make a separate ticket?) is that 4xx.html and 5xx.html pages don't include `this.store` when *intentionally* running `if (process.browser) { ... }`.

**Comment by Rich-Harris** [comment]
The problem here isn't that `preload` is running twice, it's that Sapper is trying to render the wrong page. If `preload` results in `this.error` on the server, then Sapper should simply hydrate the error page.

(Having said that, if `preload` results in different behaviour between server and client, that's arguably a bug in your app code.)

> A related bug (should I make a separate ticket?) is that 4xx.html and 5xx.html pages don't include this.store

Yeah, i just raised https://github.com/sveltejs/sapper/issues/293

**Comment by arxpoetica** [comment]
This one, unfortunately, is really hard to debug. I may have to dumb it down with an actual rudimentary app because in my current app the 400 `this.error` behavior is definitely not working correctly.

**Comment by Rich-Harris** [comment]
Can you try with 0.14?

**Comment by arxpoetica** [comment]
Yes, thanks -- won't get to it until tomorrow though --
